### PR TITLE
cli: status: only scan through mutable changes to find conflicts

### LINKED ISSUE #3628: jj status stalls for 10 seconds after printing output
## Description

In a large-ish [git repository](https://gitlab.alpinelinux.org/alpine/aports) (20k files, 200k linear commits), `jj status` takes ~1 second to snapshot the working copy, then does seemingly nothing at 100% cpu for 10 seconds after printing the output.

Linux perf says 25% of cpu time is coming from miniz_oxide::inflate::core::decompress, with stacktrace:

```
jj_cli::commands::run_command                                                                                                                                            
        <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter                                                                             
        <core::iter::sources::from_fn::FromFn<F> as core::iter::traits::iterator::Iterator>::next                                                                                      
        <jj_lib::default_index::rev_walk::FilterRevWalk<W,P> as jj_lib::default_index::rev_walk::RevWalk<I>>::next                                                                     
        jj_lib::default_index::revset_engine::build_predicate_fn::_$u7b$$u7b$closure$u7d$$u7d$::hef4c6357b53a6f23                                                                      
        jj_lib::store::Store::get_commit                                                                                                                                     
        _$LT$jj_lib..git_backend..GitBackend$u20$as$u20$jj_lib..backend..Backend$GT$::read_commit::_$u7b$$u7b$closure$u7d$$u7d$::hf7175803741d19ab                               
        _$LT$jj_lib..git_backend..GitBackend$u20$as$u20$jj_lib..backend..Backend$GT$::read_commit::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h8adf5db9f0d4bea2     
        gix::repository::object::<impl gix::types::Repository>::_find_object_inner_generated_by_gix_macro_momo                                                                  
        gix_odb::cache::impls::<impl gix_pack::find_traits::Find for gix_odb::Cache<S>>::try_find_cached                                                                       
        gix_odb::store_impls::dynamic::find::<impl gix_odb::store_impls::dynamic::Handle<S>>::try_find_cached_inner                                                                     
```

This is the FilterRevWalk used to compute [ancestor_conflicts](https://github.com/martinvonz/jj/blob/20af8c79ef9edcdcdef3bada20d1887529662c71/cli/src/commands/status.rs#L98).

Maybe jj should stop searching through ancestors when we find an immutable ancestor?

## Steps to Reproduce the Problem

```
jj git clone https://gitlab.alpinelinux.org/alpine/aports
cd aports
jj status
```

## Expected Behavior

Instantaneous command completion after printing the "Parent commit" line.

## Actual Behavior

Seconds of uselessly burning CPU.

## Specifications

- Platform: Linux, static musl release build
- Version: 0.17.0

### PR DESCRIPTION
**Pull request by edre**
Fixes #3628
Verified manually

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Comment by martinvonz** [comment]
Thank you! If we're going to get this into 0.17.1, the commit should be based on that commit and we should do a merge into `main`. I think that means I'll at least temporarily need to enable "Merge by merging" or whatever it's called in GitHub.

**Comment by martinvonz** [comment]
> Thank you! If we're going to get this into 0.17.1, the commit should be based on that commit and we should do a merge into `main`. I think that means I'll at least temporarily need to enable "Merge by merging" or whatever it's called in GitHub.

And we'll of course need to bump the version numbers too. So maybe we should create a new branch for this instead. I don't know how people normally do this with GitHub. Anyway, please don't merge this for now even if you get approval. And could you also make sure it's based on v0.17.0? Thanks

**Comment by edre** [comment]
Rebased to tag 0.17.0

**Comment by edre** [comment]
Alternatively I could merge this to main, and you could locally `jj duplicate` to whatever release branch works for you.

**Comment by martinvonz** [comment]
> Alternatively I could merge this to main, and you could locally `jj duplicate` to whatever release branch works for you.

I don't think that helps much other than for getting it more quickly to people who build from source on the `main` branch because the duplicated commit will have the same problem as your current commit has. I don't have any special permissions (though I own the repo so I could temporarily grant myself special permissions if we don't have a better solution).

File: cli/src/commands/status.rs [hunk 1]
```diff
@@ -98,6 +99,9 @@ pub(crate) fn cmd_status(
         let ancestors_conflicts = RevsetExpression::filter(RevsetFilterPredicate::HasConflict)
             .intersection(&wc_revset.ancestors())
             .minus(&wc_revset)
+            .minus(&revset_util::parse_immutable_expression(
+                &workspace_command.revset_parse_context(),
+            )?)
```
Comment (yuja) [inline cli/src/commands/status.rs:104]: nit: Since `immutable_heads()` is a user expression, `.evaluate_programmatic()` can't be used anymore.

**Comment by thoughtpolice** [comment]
> > Thank you! If we're going to get this into 0.17.1, the commit should be based on that commit and we should do a merge into `main`. I think that means I'll at least temporarily need to enable "Merge by merging" or whatever it's called in GitHub.
> 
> And we'll of course need to bump the version numbers too. So maybe we should create a new branch for this instead. I don't know how people normally do this with GitHub. Anyway, please don't merge this for now even if you get approval. And could you also make sure it's based on v0.17.0? Thanks

TBH, in this case I'd suggest that we just do "trunk based development" and just merge this, then immediately cut another release from `main` directly. We fixed barely anything in the past 7 days. I think it's completely OK to rollup the current changes on top of `0.17.0` into an `0.17.1`; there's not really much risk in that.

**Comment by martinvonz** [comment]
> TBH, in this case I'd suggest that we just do "trunk based development" and just merge this, then immediately cut another release from `main` directly. We fixed barely anything in the past 7 days. I think it's completely OK to rollup the current changes on top of `0.17.0` into an `0.17.1`; there's not really much risk in that.

We discussed this a bit on Discord. We made the changes directly on the release branch and then merged into to main branch this time. We can have a separate discussion about instead cherry-picking to the release branch in the future.

**Pull request merged**